### PR TITLE
Refactor ui tests to use @testing-library

### PR DESCRIPTION
**Pull request by jkongie**
Replaces `enzyme` with `@testing-library/react` to provide an easier to use testing library.

With this library we are now testing what the user interacts with in the UI.

This commit does not change any implementation logic and is purely focused on refactoring the tests.

### CODE REVIEW

**Comment by cl-homu** [comment]
:umbrella: The latest upstream changes (presumably #5360) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

File: package.json [hunk 1]
```diff
@@ -55,7 +55,6 @@
     "**/sha3": "^2.0.7",
     "**/solc": "^0.6.4",
     "**/source-map-support": "0.5.16",
-    "**/yargs-parser": "^15.0.1",
```
Comment (jkongie) [inline package.json:58]: No longer needed as all sub dependencies now resolve to versions which include the vulnerability patch

File: operator_ui/package.json [hunk 2]
```diff
@@ -72,10 +72,13 @@
...
     "@babel/preset-react": "^7.12.5",
     "@babel/preset-typescript": "^7.9.0",
+    "@testing-library/dom": "8.10.1",
+    "@testing-library/jest-dom": "^5.15.0",
+    "@testing-library/react": "^12.1.2",
+    "@testing-library/user-event": "^13.5.0",
     "@types/classnames": "^2.3.1",
     "@types/cookie": "^0.4.1",
     "@types/d3": "^6.2.0",
-    "@types/enzyme": "^3.10.9",
```
Comment (DeividasK) [inline operator_ui/package.json:78]: Should also remove
```
    "enzyme": "^3.11.0",
    "enzyme-adapter-react-16": "^1.15.5",
```

File: operator_ui/src/PrivateRoute.test.js [hunk 3]
```diff
@@ -24,18 +27,17 @@ const mountAuthenticatedApp = (store) =>
...
 
 describe('PrivateRoute', () => {
-  it('redirects when user is NOT autheticated', () => {
+  it('redirects when user is NOT autheticated', async () => {
     const state = { authentication: { allowed: false } }
-    const wrapper = mountAuthenticatedApp(mockStore(state))
-    expect(wrapper.find(AuthenticatedApp).props().location.pathname).toBe(
-      '/signin',
-    )
+    mountAuthenticatedApp(mockStore(state))
```
Comment (DeividasK) [inline operator_ui/src/PrivateRoute.test.js:32]:
```suggestion
    const { getByText } = mountAuthenticatedApp(mockStore(state))
```
1. I think `getBy*` should be preferred if we aren't doing async stuff (looks like that's the case here).
2. `render` returns methods that can be used. I think that should be preferred as `screen` uses a global.

File: operator_ui/src/components/BaseLink.test.tsx [hunk 4]
```diff
@@ -1,29 +1,41 @@
...
+  render(<MemoryRouter>{link}</MemoryRouter>)
+
 describe('components/BaseLink', () => {
   it('renders an anchor', () => {
-    const component = render(<BaseLink href="/foo">My Link</BaseLink>)
-    expect(component.text()).toContain('My Link')
-    expect(component.prop('href')).toEqual('/foo')
+    renderBaseLink(<BaseLink href="/foo">My Link</BaseLink>)
+
+    expect(queryByText('My Link')).toBeInTheDocument()
```
Comment (DeividasK) [inline operator_ui/src/components/BaseLink.test.tsx:15]: Should be `getByText` if we don't expect it to be `undefined`. This is a good cheat sheet - https://testing-library.com/docs/dom-testing-library/cheatsheet/#queries

File: operator_ui/src/components/BaseLink.test.tsx [hunk 5]
```diff
@@ -1,29 +1,41 @@
...
+    expect(queryByText('My Link')?.closest('a')).toHaveAttribute('href', '/foo')
```
Comment (DeividasK) [inline operator_ui/src/components/BaseLink.test.tsx:16]: Preference would be to use `getByRole('link')`

File: operator_ui/src/pages/Bridges/Index.test.js [hunk 6]
```diff
@@ -1,76 +1,88 @@
...
+
     global.fetch.getOnce(globPath('/v2/bridge_types'), bridgesResponse)
 
-    const wrapper = mountIndex()
+    renderWithRouter(<Index pageSize={1} />)
 
-    await syncFetch(wrapper)
-    expect(wrapper.text()).toContain('reggaeIsntThatGood')
-    expect(wrapper.text()).toContain('butbobistho.com')
+    await waitFor(() => getByRole('heading'))
```
Comment (DeividasK) [inline operator_ui/src/pages/Bridges/Index.test.js:31]:
```suggestion
    await findByRole('heading')
```
is equivalent and I think should be preferred

File: operator_ui/src/PrivateRoute.test.js (hunk 3 above)
Comment (DeividasK) [reply operator_ui/src/PrivateRoute.test.js:32]: Discussed in Slack and #2 should be ignored. Seems like using `screen` is best practice https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen

File: operator_ui/src/pages/Jobs/New.test.tsx [hunk 7]
```diff
@@ -47,119 +40,72 @@ describe('pages/Jobs/New', () => {
...
-      },
-    )
-    jest.useFakeTimers()
-    fillTextarea(wrapper, jobSpec1)
-    act(() => {
-      jest.runAllTimers()
-    })
-    wrapper.update()
+    // const jobSpec2 =
+    //   'observationSource = """ ds [type=ds]; ds_parse [type=ds_parse]; ds_multiply [type=ds_multiply]; """'
```
Comment (DeividasK) [inline operator_ui/src/pages/Jobs/New.test.tsx:63]: Shouldn't have commented out code

File: operator_ui/src/pages/Jobs/New.test.tsx [hunk 8]
```diff
@@ -47,119 +40,72 @@ describe('pages/Jobs/New', () => {
...
+    expect(await findByText('ds', {}, { timeout: 1000 })).toBeInTheDocument()
+
+    // This tests the update of the textbox but it doesn't currently work
+    // userEvent.paste(getByRole('textbox', { name: /TOML blob/i }), jobSpec2)
+
+    // expect(
+    //   await findByText('ds_multiple', {}, { timeout: 1000 }),
+    // ).toBeInTheDocument()
+    // expect(await findByText('ds_parse')).toBeInTheDocument()
+    // expect(await findByText('ds')).toBeInTheDocument()
```
Comment (DeividasK) [inline operator_ui/src/pages/Jobs/New.test.tsx:84]: Shouldn't have commented out code. Also, why doesn't it work? :slightly_smiling_face:

File: operator_ui/src/pages/Jobs/New.tsx [hunk 9]
```diff
@@ -156,10 +156,8 @@ export const New = ({
             setServerErrorMsg(error.toString())
           }
 
-          setValid(false)
-        })
-        .finally(() => {
           setLoading(false)
+          setValid(false)
```
Comment (DeividasK) [inline operator_ui/src/pages/Jobs/New.tsx:160]: Why this change? It seems like only doing `setLoading(false)` on error would not disable loading state on success?

File: operator_ui/src/pages/Notifications.test.js [hunk 10]
```diff
@@ -2,15 +2,17 @@
...
+import { render, screen } from '@testing-library/react'
+
+const { queryByText } = screen
 
 const classes = {}
 const mockStore = configureStore()
 
 const mountNotifications = (store) =>
-  mount(
+  render(
```
Comment (DeividasK) [inline operator_ui/src/pages/Notifications.test.js:15]: This could be trimmed if using `renderWithRouter`?

File: operator_ui/support/test-utils.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+const renderWithRouter = (
+  ui: React.ReactElement,
+  { initialEntries }: RenderWithRouterProps = { initialEntries: ['/'] },
+  options?: Omit<RenderOptions, 'wrapper'>,
+): RenderResult => {
+  return {
+    ...customRender(
+      <MemoryRouter initialEntries={initialEntries}>
+        {/* Notifications are added here because they can display a link */}
+        <Notifications />
```
Comment (DeividasK) [inline operator_ui/support/test-utils.tsx:46]: Hm, given some of your comments above, I wonder if it's worth bundling `Notifications` here. It could be added only if/when needed in tests.

File: operator_ui/src/pages/Jobs/New.test.tsx (hunk 8 above)
Comment (jkongie) [reply operator_ui/src/pages/Jobs/New.test.tsx:84]: Ah I spent too long on this one and was meant to circle back to it but I totally blanked on it. I've fixed it now

File: operator_ui/src/pages/Jobs/New.test.tsx (hunk 7 above)
Comment (jkongie) [reply operator_ui/src/pages/Jobs/New.test.tsx:63]: This has been uncommented now because it is used in the fixed test

File: operator_ui/src/pages/Jobs/New.tsx (hunk 9 above)
Comment (jkongie) [reply operator_ui/src/pages/Jobs/New.tsx:160]: This was part of my debugging of tests that I was meant to come back to. Thanks for picking that up

File: operator_ui/src/pages/Notifications.test.js (hunk 10 above)
Comment (jkongie) [reply operator_ui/src/pages/Notifications.test.js:15]: This needs to mock a store with the notification states, which you can't do with the custom render.

I'm planning to replace these types notifications anyway, so it will be removed in the near future.

File: operator_ui/support/test-utils.tsx (hunk 11 above)
Comment (jkongie) [reply operator_ui/support/test-utils.tsx]: Done!

**Review by DeividasK — APPROVED** [review]
